Refactor power distributor's `Result`s

leandro-lucarella-frequenz opened the pull request:
We make `Result` and its subclasses dataclasses so they have a proper `__str__()` method that shows the details.

We also make some property renaming to make it more consistent and improve documentation.

Fixes #275.

llucax committed:
Rename properties of power distributor results

Rename properties of Success and PartialFailure to make them more
consistent.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Improve documentation for power distributor's Request

Also improve the imports and type annotations.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Make power distributor's Results dataclasses

The Result class is an ABC but without any abtractmethods, a lot of
boilerplate can be removed by converting it to a dataclass and the same
for all subclasses.

This also uses mixins to reduce code duplication even further and making
sure result classes use consistent property names where it makes sense.

We still need to keep `Result` a class, but after the minimum supported
Python version is 3.10+ we can make `Result` a type union, so there is
not way to instantiate if by mistake, but still we can use
`isinstance(result, Success)` for example to what concrete type a result
it is, or even better use the `match` syntax.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
Please have a look at the commit messages for more info.

shsms commented on src/frequenz/sdk/actor/power_distributing/request.py:
```diff
@@ -2,22 +2,30 @@
-    # In which batteries the power should be set
-    batteries: Set[int]
-    # Timeout for the server to respond on the requests.
+    """The amount of power to be set."""
```
What do you think about:
``` suggestion
    """The requested power in watts."""

leandro-lucarella-frequenz replied:
Thanks, I didn't know the unit and thought of finding out later and then forgot!

shsms commented on src/frequenz/sdk/actor/power_distributing/request.py:
```diff
@@ -2,22 +2,30 @@
+    """The amount of power to be set."""
+
+    batteries: set[int]
+    """The set of batteries to use when requesting the power to be set."""
```
``` suggestion
    """The component ids of the batteries to be used for this request."""

shsms commented on src/frequenz/sdk/actor/power_distributing/request.py:
```diff
@@ -2,22 +2,30 @@
     adjust_power: bool = True
+    """Whether to adjust the power to match the bounds.
+
+    If `True`, the power will be adjusted (lowered) to match the bounds, so
+    only the decreased power will be set.
```
Maybe reduced works better?

``` suggestion
    If `True`, the power will be adjusted (lowered) to match the bounds, so
    only the reduced power will be set.

shsms commented on src/frequenz/sdk/actor/power_distributing/result.py:
```diff
@@ -3,122 +3,92 @@
+
+
+@dataclasses.dataclass
+class Error(Result, _BaseResultMixin):
```
`Result` already derives from `_BaseResultMixin`,  do we need to derive from it separately also?

leandro-lucarella-frequenz replied:
I guess not, it was probably an artifact from my attempt to use `Result = Success | ...`. I can remove it for now.

leandro-lucarella-frequenz replied:
Done and updated the PR.

shsms commented:
Btw,  weren't you also going to add a `__str__` or `__repr__` method to these types, so we get better logs?

llucax committed:
Improve some documentation comments

Based on the review comments.

Co-authored-by: Sahas Subramanian <sahas.subramanian@proton.me>
Co-authored-by: Daniel Zullo <daniel.zullo@frequenz.com>
Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
> Btw, weren't you also going to add a __str__ or __repr__ method to these types, so we get better logs?

dataclass takes care of it :sunglasses: 

```pycon
>>> from frequenz.sdk.actor.power_distributing.result import OutOfBound, Success
>>> oob = OutOfBound('test', 'test') # We don't need to care about types here :D
>>> str(oob)
"OutOfBound(request='test', bound='test')"
>>> repr(oob)
"OutOfBound(request='test', bound='test')"
>>> suc = Success(1, 2, 3, 4)
>>> str(suc)
'Success(succeeded_power=1, succeeded_batteries=2, excess_power=3, request=4)'
>>> repr(suc)
'Success(succeeded_power=1, succeeded_batteries=2, excess_power=3, request=4)'
```

llucax committed:
Create power distributor Results using keyword arguments

We can't enforce this for now as it is not supported in Pyton 3.8, but
ideally we should create power distributor Results using keyword
arguments to avoid mistakes and make it more clear what's what.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
I just realized that this is wrong:

```pycon
>>> str(suc)
'Success(succeeded_power=1, succeeded_batteries=2, excess_power=3, request=4)'
```

`request` should be the first argument, not the fourth. Test passed anyway because the only 2 wrong were `Success` and `PartialFailure` but both are constructed using keyword arguments anyway.

I actually wanted to make the result dataclasses `kw_only` but sadly it is not supported in Python 3.8, so we'll have to wait for that.

Anyway, I updated the PR to always use keyword arguments for all classes, just in case, and to fix the attribute order.

leandro-lucarella-frequenz commented:
Also as an extra clarification, this problem would have been caught by the type system if `Success` and `PartialFailure` were not constructed using keyword arguments.

daniel-zullo-frequenz commented on src/frequenz/sdk/actor/power_distributing/request.py:
```diff
@@ -2,22 +2,30 @@
+    only the reduced power will be set.
+
+    If `False` and the power is outside the batteries' bounds, the request will
+    fail and replied to with an `OutOfBound` result.
```
maybe `reply` instead of replied

leandro-lucarella-frequenz replied:
I don't think using `reply` is correct here. I think I'm missing a "be": "the request will fail and be replied to with an `OutOfBound` result."

But maybe a native English speaker can solve this mistery? :D

daniel-zullo-frequenz replied:
yeah `be replied` sounds right to me

daniel-zullo-frequenz commented on src/frequenz/sdk/actor/power_distributing/result.py:
```diff
@@ -3,122 +3,100 @@
 
+@dataclasses.dataclass
+class Success(_BaseSuccessMixin, Result):  # Order matters here. See above.
+    """Result returned when setting the power succeed for all batteries."""
```
```suggestion
    """Result returned when setting the power has succeeded for all batteries."""
```

leandro-lucarella-frequenz replied:
Here I wanted to put "setting the power succeeded for all batteries."

daniel-zullo-frequenz reviewed (dismissed):
I only have a couple of optional cosmetic comments

leandro-lucarella-frequenz commented:
Updated taking into account @daniel-zullo-frequenz comments.
